README.md
# chef-skyline [![Build Status](https://secure.travis-ci.org/bflad/chef-skyline.png?branch=master)](http://travis-ci.org/bflad/chef-skyline)

## Description

Installs/configures [Skyline](https://github.com/etsy/skyline).

## Requirements

### Platforms

* CentOS 6
* Fedora 17, 18, 19
* Red Hat Enterprise Linux 6
* Ubuntu 12.04, 12.10, 13.04

### Cookbooks

[Opscode Cookbooks](https://github.com/opscode-cookbooks/)

* [git](https://github.com/opscode-cookbooks/git)
* [python](https://github.com/opscode-cookbooks/python)

Third-Party Cookbooks

* [redis](https://github.com/miah/chef-redis)
* [scipy](https://github.com/bflad/chef-scipy)

## Attributes

These attributes are under the `node['skyline']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
full_duration | See templates/default/settings.py.erb | Fixnum | 86400
full_namespace | See templates/default/settings.py.erb | String | "metrics."
graphite_host | See templates/default/settings.py.erb | String | http://graphite.example.com
group | Skyline group | String | `node['skyline']['user']`
log_path | See templates/default/settings.py.erb | String | /var/log/skyline
mini_duration | See templates/default/settings.py.erb | Fixnum | 3600
mini_namespace | See templates/default/settings.py.erb | String | "mini."
oculus_host | See templates/default/settings.py.erb | String | http://oculus.example.com
pid_path | See templates/default/settings.py.erb | String | /var/run/skyline
pip_packages | PIP packages for installation | Array of Strings | See attributes/default.rb
redis_socket_path | See templates/default/settings.py.erb | String | /tmp/redis.sock
user | Skyline user | String | skyline

### Analyzer Attributes

These attributes are under the `node['skyline']['analyzer']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
algorithms | See templates/default/settings.py.erb | Array of Strings | See attributes/default.rb
analyzer_processes | See templates/default/settings.py.erb | Fixnum | 5
anomaly_dump | See templates/default/settings.py.erb | String | webapp/static/dump/anomalies.json
canary_metric | See templates/default/settings.py.erb | String | statsd.numStats
consensus | See templates/default/settings.py.erb | Fixnum | 5
min_tolerable_length | See templates/default/settings.py.erb | Fixnum | 1
max_tolerable_boredom | See templates/default/settings.py.erb | Fixnum | 100
stale_period | See templates/default/settings.py.erb | Fixnum | 500

### Horizon Attributes

These attributes are under the `node['skyline']['horizon']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
chunk_size | See templates/default/settings.py.erb | Fixnum | 10
horizon_ip | See templates/default/settings.py.erb | String | nil
max_queue_size | See templates/default/settings.py.erb | Fixnum | 500
max_resolution | See templates/default/settings.py.erb | Fixnum | 1000
pickle_port | See templates/default/settings.py.erb | Fixnum | 2024
roomba_processes | See templates/default/settings.py.erb | Fixnum | 1
skip_list | See templates/default/settings.py.erb | Array of Strings | []
udp_port | See templates/default/settings.py.erb | Fixnum | 2025
worker_processes | See templates/default/settings.py.erb | Fixnum | 2

### Webapp Attributes

These attributes are under the `node['skyline']['webapp']` namespace.

Attribute | Description | Type | Default
----------|-------------|------|--------
webapp_ip | See templates/default/settings.py.erb | String | 127.0.0.1
webapp_port | See templates/default/settings.py.erb | Fixnum | 1500

## Recipes

* `recipe[skyline]` default installation/configuration of Skyline
* `recipe[skyline::init]` symlinks Skyline init scripts and starts/enables services
* `recipe[skyline::redis]` configures Redis for Skyline from template
* `recipe[skyline::source]` installs Skyline via source

## Usage

### Skyline Installation

* Add `recipe[skyline]` to your node's run list.

## Testing and Development

Here's how you can quickly get testing or developing against the cookbook thanks to [Vagrant](http://vagrantup.com/) and [Berkshelf](http://berkshelf.com/).

    git clone git://github.com/bflad/chef-skyline.git
    cd chef-skyline
    vagrant plugin install vagrant-berkshelf
    vagrant plugin install vagrant-omnibus
    vagrant up BOX # BOX being centos6, ubuntu1204, ubuntu1210, or ubuntu1304

You can then SSH into the running VM using the `vagrant ssh` command.
The VM can easily be stopped and deleted with the `vagrant destroy`
command. Please see the official [Vagrant documentation](http://vagrantup.com/v1/docs/commands.html)
for a more in depth explanation of available commands.

## Contributing

Please use standard Github issues/pull requests and if possible, in combination with testing on the Vagrant boxes.

## License and Authors

See [LICENSE](LICENSE) file for more information.

* Author:: Brian Flad (<bflad417@gmail.com>)
* Copyright:: 2013 Brian Flad


CHANGELOG.md
## 0.3.0

* Add unittest2 and mock to default pip packages (from Skyline requirements.txt)

## 0.2.0

* Add `node['skyline']['horizon']['horizon_ip']` attribute

## 0.1.5

* Add default packages for install, starting with netcat which is used by skyline source code

## 0.1.4

* Set redis logfile attribute if redis source install and left defaulted to "stdout"

## 0.1.3

* Always create redis log directory for redis source installs

## 0.1.2

* Use "redis" as instance name in init script for redis source installs

## 0.1.1

* Additional platforms for source installation in redis recipe (fixes FC024 as well)

## 0.1.0

* Initial release


recipes/redis.rb
#
# Commented config settings are those missing from redis cookbook.
# Unfortunately, the unixsocket config is required, so all this manual work
# needs to be done below rather than using redis::server redis_instance. :(
#

node.set['redis']['config']['activerehashing'] = true
node.set['redis']['config']['appendfsync'] = "no"
#node.set['redis']['config']['auto_aof_rewrite_min_size'] = "64mb"
#node.set['redis']['config']['auto_aof_rewrite_percentage'] = 100
#node.set['redis']['config']['client_output_buffer_limit'] = [
#  'normal 0 0 0',
#  'slave 0 0 0',
#  'pubsub 32mb 8mb 60'
#]
node.set['redis']['config']['configure_no_appendfsync_on_rewrite'] = true
node.set['redis']['config']['configure_slowlog'] = true
#node.set['redis']['config']['lua_time_limit'] = 5000
node.set['redis']['config']['no_appendfsync_on_rewrite'] = true
node.set['redis']['config']['slowlog_max_len'] = 128
node.set['redis']['config']['timeout'] = 0
#node.set['redis']['config']['unixsocket'] = "/tmp/redis.sock"
#node.set['redis']['config']['unixsocketperms'] = 777

case node['platform_family']
when "debian"
  if node['platform'] == "debian"
    node.set['redis']['install_type'] = "source"
  elsif node['platform_version'].to_i <= 12
    node.set['redis']['install_type'] = "source"
  end
when "fedora"
  node.set['redis']['install_type'] = "source" if node['platform_version'].to_i <= 16 
when "rhel"
  if node['platform'] == "amazon"
    node.set['redis']['install_type'] = "source"
  elsif node['platform_version'].to_i <= 6
    node.set['redis']['install_type'] = "source"
  end
end

include_recipe "redis::install"

if node['redis']['install_type'] == "source"
  node.set['redis']['config']['logfile'] = "/var/log/redis/redis.log" if node['redis']['config']['logfile'] == "stdout"
  directory ::File.dirname(node['redis']['config']['logfile']) do
    owner node['redis']['user']
    group node['redis']['group']
    mode 00755
  end

  directory node['redis']['conf_dir'] do
    owner "root"
    group "root"
    mode 00755
  end

  directory node['redis']['config']['dir'] do
    owner node['redis']['user']
    group node['redis']['group']
    mode 00755
  end

  directory ::File.dirname(node['redis']['config']['pidfile']) do
    owner node['redis']['user']
    group node['redis']['group']
    mode 00755
  end

  template "/etc/init.d/redis-server" do
    source "redis_init.erb"
    cookbook "redis"
    owner "root"
    group "root"
    mode 00755
    variables(
      :conf_dir => node['redis']['conf_dir'],
      :logfile => node['redis']['config']['logfile'],
      :name => "redis",
      :pidfile => node['redis']['config']['pidfile'],
      :port => node['redis']['config']['port'],
      :user => node['redis']['user']
    )
    notifies :restart, "service[redis-server]"
  end
end

template "#{node['redis']['conf_dir']}/redis.conf" do
  source "redis-server.conf.erb"
  owner "root"
  group "root"
  mode 00644
  notifies :restart, "service[redis-server]"
end

service "redis-server" do
  supports :status => true, :restart => true, :reload => true
  action [ :enable, :start ]
end



attributes/default.rb

case node['platform_family']
when "debian"
  default['skyline']['packages'] = %w{
    netcat
  }
when "rhel"
  default['skyline']['packages'] = %w{
    nc
  }
else
  default['skyline']['packages'] = %w{}
end

default['skyline']['pip_packages'] = %w{
  redis
  hiredis
  python-daemon
  flask
  simplejson
  unittest2
  mock
  patsy
  statsmodels
  msgpack-python
}

default['skyline']['user'] = "skyline"
default['skyline']['group'] = node['skyline']['user']

# Shared Settings

default['skyline']['full_duration'] = 86400
default['skyline']['full_namespace'] = "metrics."
default['skyline']['graphite_host'] = "http://graphite.example.com"
default['skyline']['install_dir'] = "/opt/skyline"
default['skyline']['log_path'] = "/var/log/skyline"
default['skyline']['mini_duration'] = 3600
default['skyline']['mini_namespace'] = "mini."
default['skyline']['oculus_host'] = "http://oculus.example.com"
default['skyline']['pid_path'] = "/var/run/skyline"
default['skyline']['redis_socket_path'] = node['redis']['config']['unixsocket']

# Analyzer Settings

default['skyline']['analyzer']['algorithms'] = %w{
  first_hour_average
  mean_subtraction_cumulation
  simple_stddev_from_moving_average
  stddev_from_moving_average
  least_squares
  grubbs
  histogram_bins
}
default['skyline']['analyzer']['analyzer_processes'] = 5
default['skyline']['analyzer']['anomaly_dump'] = "webapp/static/dump/anomalies.json"
default['skyline']['analyzer']['canary_metric'] = "statsd.numStats"
default['skyline']['analyzer']['consensus'] = 5
default['skyline']['analyzer']['min_tolerable_length'] = 1
default['skyline']['analyzer']['max_tolerable_boredom'] = 100
default['skyline']['analyzer']['stale_period'] = 500

# Horizon Settings

default['skyline']['horizon']['chunk_size'] = 10
default['skyline']['horizon']['horizon_ip'] = nil
default['skyline']['horizon']['max_queue_size'] = 500
default['skyline']['horizon']['max_resolution'] = 1000
default['skyline']['horizon']['pickle_port'] = 2024
default['skyline']['horizon']['roomba_processes'] = 1
default['skyline']['horizon']['skip_list'] = []
default['skyline']['horizon']['udp_port'] = 2025
default['skyline']['horizon']['worker_processes'] = 2

# Source Settings

default['skyline']['source']['repo'] = 'git://github.com/etsy/skyline.git'
default['skyline']['source']['reference'] = "master"

# Webapp Settings

default['skyline']['webapp']['webapp_ip'] = "127.0.0.1"
default['skyline']['webapp']['webapp_port'] = 1500
